### ISSUE TITLE
twisted.names.root.Resolver eats filedescriptors

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10648]
|<img alt="jojo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jojo reported|
|-|-|
|Trac ID|trac#970|
|Type|defect|
|Created|2005-04-08 20:36:37Z|
|Branch|https://github.com/twisted/twisted/tree/root-resolver-fd-leak-970-2|

Attachments:

* [out.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/636/636fd04ebe81c6118b43c88dba88b3af34ba4c12/1e0e364a22c69340e6f02604520fdeb7674264c0.txt) (5250 bytes) - added by jojo on 2005-04-08 20:36:34Z - 
* [patchit.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/636/636fd04ebe81c6118b43c88dba88b3af34ba4c12/9b1a6813f120278a824c383bef584fe3489a2f8f.txt) (2475 bytes) - added by cb on 2009-12-01 19:59:34Z - this solves the leaking file descriptors for me ..

<details><summary>Searchable metadata</summary>

```
trac-id__970 970
type__defect defect
reporter__jojo jojo
priority__high high
milestone__ 
branch__branches_root_resolver_fd_leak_970_2 branches/root-resolver-fd-leak-970-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1112992597000000 1112992597000000
changetime__1260501532000000 1260501532000000
version__ 
owner__ 
cc__exarkun cc__spiv cc__itamarst cc__titty cc__jojo cc__teratorn cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jojo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jojo commented|
|-|-|

```
#!html
<pre>
There seems to be a problem with Twisted Names on Windows.

In twisted/names/client.py, reactor has to be passed to
ThreadedResolver.__init__():

------------------------------------------------------------------------
Index: twisted/names/client.py
===================================================================
--- twisted/names/client.py	(Revision 13495)
+++ twisted/names/client.py	(Arbeitskopie)
@@ -359,6 +359,7 @@
 
 def createResolver(servers = None, resolvconf = None, hosts = None):
     from twisted.names import resolve, cache, root, hosts as hostsModule
+    from twisted.internet import reactor
     if platform.getType() == 'posix':
         if resolvconf is None:
             resolvconf = '/etc/resolv.conf'
@@ -369,7 +370,7 @@
     else:
         if hosts is None:
             hosts = r'c:\windows\hosts'
-        bootstrap = ThreadedResolver()
+        bootstrap = ThreadedResolver(reactor) # Pass reactor as argument!
         hostResolver = hostsModule.Resolver(hosts)
         theResolver = root.bootstrap(bootstrap)
------------------------------------------------------------------------

The following simple test stops working after a while and produces the
attached output. When using ThreadedResolver instead of
createResolver(), everything works fine.

from twisted.names.client import ThreadedResolver, createResolver
from twisted.internet import reactor

class TestResolver(object):
    def __init__(self):
	self.num_connections = 0
	self.resolver = createResolver()
        #self.resolver = ThreadedResolver(reactor)

    def resolve(self):
	self.num_connections += 1
	d = self.resolver.getHostByName("www.twistedmatrix.com")
	d.addCallback(self.gotIP)
	reactor.callLater(0.1, self.resolve)

    def gotIP(self, ip):
	self.num_connections -= 1
	print "Got IP %s. %d connections." % (ip, self.num_connections)

t = TestResolver()
t.resolve()
reactor.run()
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
First part of this is fixed and checked in at r13513 and r13514

Dunno what to do about the second part.  It seems to be a win32-only problem. 
It may even have nothing to do with t.names, I can't tell.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

This isn't win32 only.

Change 'self.resolver = ...' to  
self.resolver = twisted.names.root.Resolver(['192.228.79.201'])

This results in the following error message on linux:
```
Unhandled error in Deferred:
Traceback (most recent call last):
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/defer.py", line 229, in callback
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/defer.py", line 294, in _startRunCallbacks
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/defer.py", line 307, in _runCallbacks
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/defer.py", line 630, in gotResult
    
--- <exception caught here> ---
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/defer.py", line 606, in _deferGenerator
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/names/root.py", line 127, in discoverAuthority
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/names/root.py", line 74, in lookupNameservers
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/names/root.py", line 38, in retry
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/names/dns.py", line 1126, in query
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/names/dns.py", line 1084, in startListening
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/posixbase.py", line 307, in listenUDP
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/udp.py", line 84, in startListening
    
  File "/scratch1/ralf/bbot/extern/Twisted-trunk/twisted/internet/udp.py", line 92, in _bindSocket
    
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

While I'm here:
Shouldn't the twisted.names.root.bootstrap mechanism be replaced by a static list of IP addresses?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Ah.  It's pretty clear what's going on here now.  root.Resolver needs to clean up the resolver it creates for each lookup it does.   I'll take care of this shortly.

Regarding bootstrap, what would the advantage of this be?  The list would just have to be maintained and updated, and old installs of Twisted would randomly break over time.

What it should actually be replaced (or probably just augmented) with is a platform-specific mechanism for determining the system's __actual__ configured nameservers.

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

The advantage is that one can use root.Resolver without first using dns resolving (which might not work).

The platform specific mechanism is working for unix, still root.Resolver isn't used (this is fine). In my opinion this has nothing to do with being able to determine the configured nameservers.

Maybe, it should just be rewritten with a static list of IPs and then using root.Resolver([ips]) to do the name lookups for the root servers. This would also be real bootstrapping :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Certainly this should still be fixed, but I'm not likely to get to it in the foreseeable future.  If someone would like to supply a patch with unit tests, I'll gladly review it, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

[#3982](https://github.com/twisted/twisted/issues/3982) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

[#4141](https://github.com/twisted/twisted/issues/4141) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|cb commented|
|-|-|

I think this bug has a pretty high priority ..

It shows that might be a design-bug in the twisted.named.root.Resolver()

being able to bootstrap a recursive caching Resolver is very important

at the Moment the root.Resolver() is pretty much useless and should be marked broken

yours
  Christian

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|cb commented|
|-|-|

Replying to [cb](#note_13):
please also have look at [#4141](https://github.com/twisted/twisted/issues/4141)

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|cb commented|
|-|-|

i fixed the leaking resolver ports for my use case ..
.. though it seems to have problems when the reactor is not running?

please see the attached patchit.txt

yours
 Christian Bahls

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Needs unit tests badly.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [27646]) Branching to 'root-resolver-fd-leak-970'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [27694]) Branching to 'root-resolver-fd-leak-970-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|TimAllen set owner to @exarkun|
|-|-|

OK, this is my first attempt at a code-review. Hope this works! 
 
 1. `t/n/client.py`, line 2: File has been edited, but copyright date not bumped. 
 1. `t/n/client.py`, lines 33-35: You've deleted the `_errormap` variable from `Resolver`, but the exceptions it contained are still being imported even though they're not used. 
 1. It's not immediately obvious why `t.n.dns.RRHeader` has learned how to compare itself to other instances. I mean, it's a useful-looking change, but is it necessary   for the rest of the code in this patch to work?
 1. `t/n/root.py`, line 121: The list of timeouts `(1, 3, 11, 45)` appears in nearly a dozen places through the Twisted codebase, and three times in `root.py` alone.     Does this sequence come from an RFC or something? Should it be a module-level constant?
 1. `t/n/root.py`, in `Resolver`: This code is very confused about whether it wants a single server to query, or a list of them. `__init__()` takes a list, which         `_lookup()` passes to `_discoverAuthority()`, which grabs the server at the head of the list and throws the tail away. The head gets passed to `_query()` which wraps it  back into a list and passes it to `client.Resolver()`. Later on in `_discoveredAuthority()`, we take a list of potential servers in `traps`, look up only the first one,  and wrap it back into a list to pass to `_discoverAuthority()` (which will unpack it so `_query()` can re-pack it). Since this class is given a list and passes a list to the underlying utility functions, they might as well stay lists all along.
 1. `t/n/root.py`, line 137: The docstring says `timeout` should be a list of ints, but everywhere that this code supplies a timeout, it's a tuple of ints.
 1. `t/n/root.py`, line 183: There's a comment that says "Check class too!" If this is just a matter of adding "and answer.class == query.class", you might as well do it.
 1. `t/n/root.py`, line 206: If we get a response with a list of authoritative name servers and addresses for *some* of them, is it OK to just forward the query to the   servers we have addresses for and ignore the others? If we're ignoring information, doesn't that make things less reliable?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [27697]) Update copyright date; remove unused imports

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [27698]) Check the RR class before considering it an answer to our query

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [27699]) Pass around just one server address while recursing, since only server is ever actually used

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [27700]) timeouts are typically tuples, not lists

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [27701]) A test for the behavior of a lookup method when there is a CNAME for the QNAME in the response, but no record matching the query type

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [27702]) A test for cname cycles

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [27703]) Follow CNAMEs within a single response, detecting cycles.

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun removed owner|
|-|-|

Thanks for the in-depth review.

  1. Date updated
  1. Unused imports deleted
  1. This was to ease testing.  In a couple places the tests want to verify that the right `RRHeader` is returned.
  1. It's somewhat recommended by some standard, I don't remember which.  I think it's an API mistake, but it's hard to fix at this point (I think there's room for a higher-level, nicer API, where perhaps it can be fixed).  Since most of the occurrences in root.py are in deprecated code, I'm not too bothered.  The net count will go down once that code is deleted.
  1. Indeed.  It keeps lists around in more places now.
  1. Fixed.
  1. Added a test for that and implemented the class check.
  1. That's a pretty good question.  I'm not sure what the current best practice is for this.  Supposedly, every server included in that response is supposed to be able to answer the query.  Of course, in practice, a server might be down, overloaded, misconfigured, etc.  It might be sensible to make sure all servers are tried.   I think this could easily be a separate unit of work, too, though.  And it's pretty reliable as-is.

Beyond what was mentioned in the review, I also implemented some more CNAME logic (following them with an additional query when their target is not included in the response, detecting cycles).  I'm not all that happy with how long `_discoveredAuthority` is now.  The processing logic is somewhat complex, and the implementation reflects this.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TimAllen set owner to TimAllen|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_32"></a>|TimAllen set owner to @exarkun|
|-|-|

I did a brief search for the sequence "1, 3, 11, 45", and the only hits are
for various copies of the Twisted source across the internet, and a Japanese
PDF about DNSSEC. Oh, and it's the first few terms of the sequence of Super
Catalan Numbers (they fight crime!). I think it'd be good to have a comment mentioning that these numbers are arbitrary rather than being
fundamentally important parts of some DNS RFC, just to ward off cargo-cult
programmers (it may be too late for Japanese DNSSEC researchers).

 1. `t/n/root.py`, in `findAnswerOrCName()`: You say "`if record.name == name`". Since the `records` variable was populated by grabbing the `.name` attribute from each record, surely that condition should always be true.
 1. `t/n/root.py`, line 207 in `findAnswerOrCName()`: It might be worth pointing out in a comment that a given DNS name should only ever have one CNAME, so it doesn't matter that this method always chooses the last CNAME.
 1. `t/n/root.py`, in `_discoverAuthority()`: I agree, this method is getting unwieldly and long, especially the "while True" loop is a bit mind-bending. On the other hand, I don't have any concrete suggestions on how to improve it either.
 1. `t/n/root.py`, line 266 in `_discoverAuthority()`: If the reply included no `hints`, this code assumes it must have included at least one entry in `traps`, which seems a bold assertion (there's got to be *some* sufficiently misconfigured DNS server out there somewhere). In this circumstance, raising some kind of 'DNS server is stupid' error would be more helpful than `IndexError: list index out of range`. Likewise for assuming that the first entry in `traps` resolves to at least one sensible IP address.
 1. `t/n/t/test_rootresolve.py`, in `_getResolver` docstring: The "@param" line needs an extra space of indent.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun commented|
|-|-|

(In [27718]) Remove redundant name check

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun commented|
|-|-|

(In [27720]) Comment about the cname handling in findAnswerOrCName

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

(In [27722]) Handle the case where a response has no answers and is not a delegation

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [27723]) Add a test making sure nameserver lookup errors propagate

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun commented|
|-|-|

(In [27724]) Add a test making sure another kind of nameserver lookup error propagates

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun commented|
|-|-|

(In [27725]) Fix whitespace in _getResolver docstring

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun commented|
|-|-|

(In [27726]) comment about why we use 1, 3, 11, 45

refs [#10648](https://github.com/twisted/twisted/issues/10648)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun removed owner|
|-|-|

Thanks!  All points addressed.

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_41"></a>|TimAllen set owner to @exarkun|
|-|-|

Looks good to me! Ready to merge, as far as I can see.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun set status to `closed`|
|-|-|

(In [27740]) Merge root-resolver-fd-leak-970-2

Author: exarkun
Reviewer: TimAllen
Fixes: [#10648](https://github.com/twisted/twisted/issues/10648)

Fix `twisted.names.root.Resolver` so that it does not leak several file descriptors
(udp ports) for each lookup it performs.  Also re-introduce tests for this code and
deprecate several of the previous APIs which were really only implementation details,
not functionality intended to be made public.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
